### PR TITLE
Use GroupConvolutionBackpropData v1 in ONNX ConvTranspose.

### PR DESCRIPTION
**Pull request by arogowie-intel**
Short summary of changes:
* `GroupConvTranspose` fused operator was removed as it was used only by ONNX Importer.
* Add helper functions for infering output spatial dimensions or auto padding values for Convolution Bakcprop
* `v1::[Group]ConvolutionBackpropData`:
  * Added decomposition implementation only in (`Group~` operator)
  * Updated validation logic 
  * Added custom `is_dynamic` overload with more restrictive check
  * Added new constructor not requiring `pads_begin` and `pads_end` attributes.
  * Updated the constructors doc with inputs shapes.
  * `get_output_shape` returns values only for spatial dimensions.
* Updated Upgrade/downgrade passes for `[Group]ConvolutionBackpropData`
* Updated UT and moved tests for `GroupConvTranspose` to `GroupConvolutionBackpropData`

### CODE REVIEW

File: src/ngraph/op/convolution.cpp [hunk 1]
```diff
@@ -207,12 +207,12 @@ op::v1::ConvolutionBackpropData::ConvolutionBackpropData(const Output<Node>& dat
 const PartialShape op::v1::ConvolutionBackpropData::get_output_shape() const
 {
     PartialShape shape{vector<Dimension>(m_strides.size() + 2)};
-    auto data_pshape = get_input_partial_shape(0);
+    auto data_pshape = input(0).get_partial_shape();
```
Comment (diyessi) [inline src/ngraph/op/convolution.cpp]: I decided `get_input_partial_shape(i)` is too convenient and have a PR that un-deprecates them, so its up to you if you want these changes or not.

File: src/ngraph/pass/opset0_downgrade.cpp [hunk 2]
```diff
@@ -178,26 +178,13 @@ namespace
...
-        bool is_op_valid = all_of(
-            output_padding.begin(), output_padding.end(), [](size_t value) { return value == 0; });
+        const auto& strides = node->get_strides();
+        const auto& dilations = node->get_dilations();
+        const auto& output_padding = node->get_output_padding();
 
-        NGRAPH_CHECK(is_op_valid,
-                     "Unable to convert ConvolutionBackpropData:v1 to ConvolutionBackpropData:v0 "
-                     "with output padding other than `0`. Node: ",
-                     *node);
```
Comment (tsocha) [inline src/ngraph/pass/opset0_downgrade.cpp]: Could you explain why output_padding verification has been removed?

File: test/opset_pass/convolution_opset_pass.cpp [hunk 3]
```diff
@@ -180,7 +183,7 @@ TEST(opset_transform, opset1_group_convolution_backprop_data_upgrade_pass)
     auto padding_end = CoordinateDiff{3};
 
     auto group_conv_backprop = make_shared<op::v0::GroupConvolutionBackpropData>(
-        data_batch_shape, filters, delta, strides, dilations, padding_begin, padding_end, 1);
+        data_batch_shape, filters, delta, strides, dilations, padding_begin, padding_end, 4);
```
Comment (tsocha) [inline test/opset_pass/convolution_opset_pass.cpp]: [NON-BLOCK]
Just wonder if using `groups` variable instead of `4` value will improve readability.

File: src/ngraph/frontend/onnx_import/op/conv_transpose.cpp [hunk 4]
```diff
@@ -82,30 +82,39 @@ namespace ngraph
...
+
                     std::shared_ptr<ngraph::Node> conv_node;
                     if (!output_shape.empty())
                     {
-                        conv_node = std::make_shared<ngraph::opset0::GroupConvolutionTranspose>(
+                        conv_node = std::make_shared<default_opset::GroupConvolutionBackpropData>(
                             data,
                             filters,
+                            default_opset::Constant::create(
+                                element::i64, Shape{output_shape.size()}, output_shape),
```
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/op/conv_transpose.cpp]: In ONNX the `output_shape` is defined also only for _spatial shapes_. Take a look at equation: 
```
output_shape[i] = stride[i] * (input_size[i] - 1) + output_padding[i] + ((kernel_shape[i] - 1) * dilations[i] + 1) - pads[start_i] - pads[end_i]
```
We have also such UT models: https://github.com/NervanaSystems/ngraph/blob/aosewski/group_conv_bprop_v1/test/models/onnx/convtranspose_output_shape.prototxt#L10-L19.
I'll add check to test whether `output_shape` size is only of _spatial shape's_ size.

File: src/ngraph/op/fused/group_conv.hpp [hunk 5]
```diff
@@ -129,16 +134,45 @@ namespace ngraph
...
+                //                             each spatial axis in the output tensor.
+                //
+                // clang-format on
+                GroupConvolutionBackpropData(const Output<Node>& data,
+                                             const Output<Node>& filter,
+                                             const Output<Node>& output_shape,
+                                             const Strides& strides,
+                                             const Strides& dilations,
+                                             const PadType& auto_pad,
+                                             const CoordinateDiff& output_padding = {});
```
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/op/fused/group_conv.hpp]: The V1 spec is actually inconsistent in this area. There is information at the beginning that:

> If `output_shape` is specified, neither `pads_begin` nor `pads_end` should be specified, but 
> `auto_pad` defines how to distribute padding amount around the tensor.

Later, we can see in both `pads_begin` and `pads_end` attributes description:
> May be omitted, in which case pads are calculated automatically.
However both are marked _required_.

IMHO those attributes shouldn't be required. Yes I agree we could change spec here. @GlebKazantaev  @ilyachur What do you guys think about this?

File: src/ngraph/op/fused/group_conv.cpp [hunk 6]
```diff
@@ -225,77 +253,145 @@ void op::v1::GroupConvolutionBackpropData::set_output_shape(const Shape& shape)
...
+            Shape output_shape = output_pshape.to_shape();
+            // extract spatial shape values
+            // output_shape.erase(std::begin(output_shape), std::next(std::begin(output_shape), 2));
+            const Shape& data_shape = data_pshape.to_shape();
+            const Shape& filters_shape = filters_pshape.to_shape();
+            const size_t num_spatial_dims = data_shape.size() - 2;
+            NODE_VALIDATION_CHECK(this,
+                                  output_shape.size() - 2 == num_spatial_dims,
+                                  "Output shape should be specified only and for "
+                                  "all spatial dimensions.");
```
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/op/fused/group_conv.cpp]: I think that both this check and the statement you mentioned are working since the `get_output_shape` inserts values for batch and channel dimensions. However I came into conclusion that the method `get_output_shape()` should return dimensions only for spatial features as it is defined in the documentation.

File: src/ngraph/op/fused/group_conv.cpp [hunk 7]
```diff
@@ -225,77 +253,145 @@ void op::v1::GroupConvolutionBackpropData::set_output_shape(const Shape& shape)
...
+        if (m_dilations.size() == 0)
+        {
+            m_dilations = conv_default_strides(this, data_pshape, filters_pshape);
+        }
+
+        const size_t num_spatial_dims = data_shape.size() - 2;
+
+        NODE_VALIDATION_CHECK(this,
+                              m_strides.size() == num_spatial_dims,
+                              "Strides should be defined for all and only spatial features.");
```
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/op/fused/group_conv.cpp]: I think that the _data_ input should define the number of spatial features and the strides size have to be compared to it. Actually If we're dealing with dynamic size of data input how could we be sure that the `strides` size is correct? After the node validation is done and we have static input data size then, that's right it's convenient to use `strides` size. But IMHO the input data size should have the priority in calculating the number of spatial features.

File: src/ngraph/op/fused/group_conv.cpp [hunk 8]
```diff
@@ -225,77 +253,145 @@ void op::v1::GroupConvolutionBackpropData::set_output_shape(const Shape& shape)
...
+                {
+                    m_pads_begin[i] = total_padding / 2;
+                    m_pads_end[i] = total_padding - m_pads_begin[i];
+                }
+                else
+                {
+                    m_pads_end[i] = total_padding / 2;
+                    m_pads_begin[i] = total_padding - m_pads_end[i];
+                }
+            }
```
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/op/fused/group_conv.cpp]: Yup :)

File: src/ngraph/op/fused/group_conv.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (tsocha) [reply src/ngraph/op/fused/group_conv.cpp]: Ou i missed this little `- 2` sorry for that(maybe I get confused by validation message).
Since you have mentioned that ONNX specify only spatial dimensions for `output_shape` we can update this method.

File: src/ngraph/builder/reshape.cpp [hunk 9]
```diff
@@ -229,7 +229,7 @@ shared_ptr<Node> builder::opset1::expand_dims(const Output<Node>& value, size_t
     return builder::opset1::reshape(value, output_shape);
 }
 
-shared_ptr<Node> builder::opset1::squeeze(const Output<Node>& value, vector<int64_t> axes)
+shared_ptr<Node> builder::opset1::squeeze(const Output<Node>& value, vector<size_t> axes)
```
Comment (ilyachur) [inline src/ngraph/builder/reshape.cpp]: Why did you change `int64_t` to `size_t`? The int64_t has a defined size, the size_t is depend on OS bits.

File: src/ngraph/op/fused/group_conv.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply src/ngraph/op/fused/group_conv.hpp]: @slyalin this is your turn

**Comment by GlebKazantaev** [comment]
@arogowie-intel do you have enough confidence that all this code with convolutions are well covered? I'm worried just because convolutions are everywhere and this PR contains a lot of changes that I'm actually can not review with 100% confidence.

File: src/ngraph/frontend/onnx_import/op/conv_transpose.cpp [hunk 10]
```diff
@@ -115,10 +138,14 @@ namespace ngraph
                     }
 
                     auto bias = inputs.at(2);
-
-                    return {std::make_shared<ngraph::opset0::Add>(
-                        conv_node,
-                        ngraph::op::make_broadcast_node(bias, conv_node->get_shape(), 1))};
+                    const Shape& new_shape = conv_node->get_shape();
+                    auto broadcasted_bias = std::make_shared<default_opset::Broadcast>(
```
[the remark this answers is not in the document]
Comment (tsocha) [reply src/ngraph/frontend/onnx_import/op/conv_transpose.cpp]: ONNX use biases with `[C]` shape so we need to broadcast it to the `output_shape` or reshape it to `[1,C,1,1]` (for 4D case) to perform an `add` operation since `numpy style` can't broadcast `[C]` into `[N, C, H, W]`.
ONNX doc:
```
B (optional) : T
    Optional 1D bias to be added to the convolution, has size of C.
```
IMHO we should adjust conv->bias fusion to support `bradcast bias` case as well as `[1, C, 1, 1]` since user can create a model with separate `convolution` and `add`(for bias) operations.

[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/op/conv_transpose.cpp]: I agree with @tsocha. NumPy broadcasting mechanism won't work for this case.

File: src/ngraph/builder/reshape.cpp (hunk 9 above)
Comment (arogowie-intel) [reply src/ngraph/builder/reshape.cpp]: Since this is just a helper function which is intended to call with already _normalized_ axes. It internally doesn't handle negative axes so in this way I enforce it's correct behavior. This is not an implementation of `v1::Squeeze` operator which you can find here: https://github.com/NervanaSystems/ngraph/blob/aosewski/group_conv_bprop_v1/src/ngraph/op/fused/squeeze.cpp#L35-L92.

File: src/ngraph/builder/reshape.cpp [hunk 11]
```diff
@@ -229,7 +229,7 @@ shared_ptr<Node> builder::opset1::expand_dims(const Output<Node>& value, size_t
     return builder::opset1::reshape(value, output_shape);
 }
 
-shared_ptr<Node> builder::opset1::squeeze(const Output<Node>& value, vector<int64_t> axes)
```
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/builder/reshape.cpp]: Since this is just a helper function which is intended to call with already normalized axes. It internally doesn't handle negative axes so in this way I enforce it's correct behavior. This is not an implementation of v1::Squeeze operator which you can find here: https://github.com/NervanaSystems/ngraph/blob/aosewski/group_conv_bprop_v1/src/ngraph/op/fused/squeeze.cpp#L35-L92.

File: src/ngraph/frontend/onnx_import/op/conv_transpose.cpp [hunk 12]
```diff
@@ -65,47 +65,70 @@ namespace ngraph
...
                     }
                     else
                     {
-                        conv_node = std::make_shared<ngraph::opset0::GroupConvolutionTranspose>(
+                        conv_node = std::make_shared<default_opset::GroupConvolutionBackpropData>(
                             data,
                             filters,
                             strides,
-                            dilations,
                             padding_below,
```
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/op/conv_transpose.cpp]: This comes from nGraph v0 operators. They used to have such nomenclature in all convolutions, ie here: https://github.com/NervanaSystems/ngraph/blob/aosewski/group_conv_bprop_v1/src/ngraph/op/convolution.hpp#L293-L294.

File: src/ngraph/op/convolution.cpp [hunk 13]
```diff
@@ -204,30 +204,35 @@ op::v1::ConvolutionBackpropData::ConvolutionBackpropData(const Output<Node>& dat
     constructor_validate_and_infer_types();
 }
 
+bool op::v1::ConvolutionBackpropData::is_dynamic() const
+{
+    bool is_dynamic = Node::is_dynamic();
+    if (get_inputs().size() == 3 && !is_dynamic)
+    {
+        return !is_type<op::Constant>(input_value(2).get_node());
```
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/op/convolution.cpp]: We don't have to check for batch and channel dims since they're already checked here: https://github.com/NervanaSystems/ngraph/blob/aosewski/group_conv_bprop_v1/src/ngraph/node.cpp#L954-L964 while data and filters inputs are checked. The third input `output_shape` must be additionally checked, since it's not sufficient to know whether it's shape is dynamic. At this moment there's no support for _dynamic_ output shape.

File: src/ngraph/op/convolution.cpp [hunk 14]
```diff
@@ -204,30 +204,35 @@ op::v1::ConvolutionBackpropData::ConvolutionBackpropData(const Output<Node>& dat
...
     {
         if (auto const_op = as_type<op::Constant>(input_value(2).get_node()))
         {
-            auto output_shape = const_op->get_shape_val();
-            // Populate spatials
-            for (int i = 0; i < output_shape.size(); ++i)
-            {
-                shape[i + 2] = output_shape[i];
-            }
+            shape = const_op->get_shape_val();
```
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/op/convolution.cpp]: This must have been changed after I reverted renaming `output_shape` input to `output_spatial_shape`. I'll update this.

File: src/ngraph/op/fused/group_conv.cpp [hunk 15]
```diff
@@ -190,30 +219,35 @@ op::v1::GroupConvolutionBackpropData::GroupConvolutionBackpropData(
     constructor_validate_and_infer_types();
 }
 
+bool op::v1::GroupConvolutionBackpropData::is_dynamic() const
+{
+    bool is_dynamic = Node::is_dynamic();
+    if (get_inputs().size() == 3 && !is_dynamic)
+    {
+        return !is_type<op::Constant>(input_value(2).get_node());
```
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/op/fused/group_conv.cpp]: As above.

**Comment by arogowie-intel** [comment]
@GlebKazantaev I understand your concerns. I tried to be as close as it is possible to the documentation. All current UT are passing so we have some confidence that this PR won't brake anything we have working until now.

File: src/ngraph/builder/reshape.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply src/ngraph/builder/reshape.cpp]: I'm not agree. Why should I care about axes normalization? As I understand builder returns V1::Squeeze op and I don't want to normalize axes. From my point of view axes normalization is only required in V1::Squeeze shape inference.
Also we have a lot of bugs in nGraph where AxisSet is used and negative values leads to undefined behavior.

File: src/ngraph/frontend/onnx_import/op/conv_transpose.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply src/ngraph/frontend/onnx_import/op/conv_transpose.cpp]: Ok, but in this place we convert ONNX Convolution to V1 nGraph Convolution and I see no reason to keep this confusing names.

File: src/ngraph/frontend/onnx_import/op/conv_transpose.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply src/ngraph/frontend/onnx_import/op/conv_transpose.cpp]: Yes in ONNX it's just [C] and in converter we need to V1::Reshape it to [1,C,1,1] (for 4D case).
And then pass this reshaped constant to V1::Add(convolution, bias_reshape).

File: src/ngraph/op/convolution.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply src/ngraph/op/convolution.cpp]: Agree. Missed this line.

File: src/ngraph/op/fused/group_conv.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (slyalin) [reply src/ngraph/op/fused/group_conv.hpp]: @arogowie-intel, the spec is not very defined regarding {Group}ConvolutionBackpropData. We need a dedicated person to fine-tune both specification and implementation. Please create MR to the specification that solves all issues that you found.

**Comment by GlebKazantaev** [comment]
@arogowie-intel do we have tests that covers conversion from ONNX to nGrpah function?

File: src/ngraph/frontend/onnx_import/op/conv_transpose.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (postrational) [reply src/ngraph/frontend/onnx_import/op/conv_transpose.cpp]: Shouldn't we adjust the conv->bias fusion to support both cases? 
We can encounter user models which use broadcast bias case instead of [1, C, 1, 1]?

[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply src/ngraph/frontend/onnx_import/op/conv_transpose.cpp]: In [N,C,H,W] cases we need to analyze tensor to detect that it's actually [1,C,1,1]. That is not hard to implement but we haven't seen such cases yet and I think won't see.
Also why I'm against [N,C,H,W] broadcasting is that if we have a sequence Conv->Add(NCHW)->BN first BN will be decomposed to linear ops and then Add(NCHW)->Mul(1,C,1,1)->Add(1,C,1,1)->Mul(1,C,1,1)->Add(1,C,1,1) sequence will be fused to the Mul(NCHW)->Add(NCHW) sequence that takes much more time in constant folding than just converting it to Mul(1,C,1,1)->Add(1,C,1,1). I can measure but we can avoid it with a simple solution: just using [1,C,1,1] for convolution biases.

**Comment by arogowie-intel** [comment]
> 
> 
> @arogowie-intel do we have tests that covers conversion from ONNX to nGrpah function?

Apart from one in our repo: https://github.com/NervanaSystems/ngraph/blob/aosewski/group_conv_bprop_v1/test/onnx/onnx_import_convpool.in.cpp#L414-L427
There are UT in ONNX repository under this path: https://github.com/onnx/onnx/tree/master/onnx/backend/test/data. If you use file search and type `convtranspose.onnx` then you should find 12 tests.

File: src/ngraph/builder/reshape.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/builder/reshape.cpp]: If you do not want to have bugs because of negative axes, then axes normalization is the way to go. I'm not sure how many nGraph ops has support for negative axis value, but a lot of ONNX operators can accept such values. As I said earlier this is a helper function which forces developer to not pass negative axis value, which would lead to errors.

File: src/ngraph/frontend/onnx_import/op/conv_transpose.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/op/conv_transpose.cpp]: Changed.

File: src/ngraph/builder/reshape.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/builder/reshape.cpp]: Moreover the axes values have to be non negative, otherwise we would get error here: https://github.com/NervanaSystems/ngraph/blob/master/src/ngraph/builder/reshape.cpp#L242. 
And currently `ngraph::Squeeze` doesn't actually support negative axes: https://github.com/NervanaSystems/ngraph/blob/master/src/ngraph/op/fused/squeeze.cpp#L47

[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply src/ngraph/builder/reshape.cpp]: Yes I do not wan to have bugs. So I've created a bug for Squeeze and Unsqueeze ops as they must support negative axes according to spec.
Also talking about bugs, how you prohibit user to pass negative values to size_t argument?

File: src/ngraph/frontend/onnx_import/op/conv_transpose.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply src/ngraph/frontend/onnx_import/op/conv_transpose.cpp]: Another most critical reason is that Broadcast brakes network reshapability. For example if you want to change input shape for example from [N,C,H,W] to [N+1,C,H,W] where N > 1, this Broadcasted Constant with [N,C,H,W] shape won't be broadcasted to [N+1,C,H,W]. That's why you need to produce Add with [1,C,1,1] Constant shape.

[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply src/ngraph/frontend/onnx_import/op/conv_transpose.cpp]: Just simply change Broadcast to V1::Reshape with [1,C,1,1] shape and you won't have this troubles.

File: src/ngraph/builder/reshape.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/builder/reshape.cpp]: Regarding the Squeeze/Unsqueeze support for negative values, that's right, they should be fixed, however they're out of the scope of this PR. The user of this _helper function_ should be a frontend/core developer.
I do not prohibit ONNX user. The ONNX model can have negative values and we simply first normalize them in ONNX Importer code.

File: src/ngraph/frontend/onnx_import/op/conv_transpose.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/op/conv_transpose.cpp]: OK, changed.
